### ISSUE TITLE
How is the "correct" way to debug the Wren when developing it?

### ISSUE THREAD

**Issue by macabeus** [munificent/wren#553]
When I was studying compilers I found the Wren, and I want to developer it, but I think that is very hard programming in C without a debugger, like LLDB.

I trying many ways to debug the Wren, but I can't find the best way to debug the Wren.

* On lldb, the console isn't type any command:
```
➜  wren git:(master) ✗ lldb wren
(lldb) target create "wren"
Current executable set to 'wren' (x86_64).
(lldb) process launch
Process 41345 launched: '/Users/macabeus/Desktop/wren/wren' (x86_64)
\\/"-
 \_/   wren v0.1.0
> var x = 1Unhandled byte: 10
> var x = 1Unhandled byte: 10
> var x = 1Unhandled byte: 10
```

* I tried many times to use Clion, but I can't attach the executable on debug

* I tried many times to use lldb.nvim plugin, [but I can't work too](https://stackoverflow.com/questions/50671977/error-when-try-to-import-lldb-in-a-python-script)

* On Xcode we have breakpoint, but the console has bug (colors and refresh isn't work).

I "fix" the console on Xcode erasing the function refreshLine and colors, but it is a workaround (#554).

Well... How to debug the Wren? C isn't my main language, sorry if I said something wrong.

**Comment by mhermier** [comment]
I don't think it is a C error, but most probably Xcode failing to emulate an ANSI console, or our test being to much optimistic.
Anyway, I think you should have a working terminal if you replace in https://github.com/munificent/wren/blob/master/src/module/repl.wren#L928:
if (Platform.isPosix && Stdin.isTerminal) {
by
if (false) {
So you get a SimpleRepl.

**Comment by macabeus** [comment]
@mhermier Thank for your reply.
I tried it, but it still didn't works on `refreshLine`. When I try to type `var x = 1` I get on console:

![image](https://user-images.githubusercontent.com/9501115/40951494-a74ca94a-684d-11e8-8b2b-55f0dc8b9c2c.png)

**Comment by mhermier** [comment]
Ok I think I see what is happening. Despite the coloring is disabled, I
think it still try to issue the colored *input help* but since ANSI term is
disabled, it fails somehow at cleaning the line.

**Comment by mhermier** [comment]
Looked at the code and indeed the problem is in refreshLine.
On Mac `\r` is used for newline, while its `\n` on linux and `\r\n` on Windows.
It seems that since Max OS Lion they went the Unix way, but maybe the issue still exists in Xcode console emulation.
Else you should disable refresh line and remove the echo disabling code `Stdin.isRaw = true`.

**Comment by Akaricchi** [comment]
>On Mac \r is used for newline

IIRC this has not been true ever since the first Mac OS X release. I'm not a mac user and can only cite this random piece of documentation: https://ftfy.readthedocs.io/en/latest/#ftfy.fixes.fix_line_breaks

**Comment by acook** [comment]
@Akaricchi is correct. Mac OS9 and before *did* use `\r` but macOS X and beyond are based on NeXTSTEP derived from BSD and thus uses the standard Unix-like newline `\n`. I have used macOS extensively and can confirm this, however I am less familiar with XCode specifically to know if they are mangling anything or using legacy (pre OSX) code for their display console.

**Comment by macabeus** [comment]
@acook @Akaricchi @mhermier I fix it on #577. Could you review it, please?

**Comment by munificent** [comment]
The short answer to your question is that I personally tend to debug using XCode, but not using the Wren REPL. Instead, I usually write a little script and have XCode run Wren on that script.

I definitely agree the REPL is pretty bad inside the XCode console right now and should be improved. I'm going to close out this issue because there's another issue to track using the REPL inside XCode.
